### PR TITLE
core: nfp: Rewrite implementation to remove direct access from the frontend

### LINKED ISSUE #8493: [Splatoon 2] Amiibo do not work correctly and cannot be saved to
issue:
in Splatoon 2, tapping an amiibo works but trying to save to an amiibo does not, restricting the ability to use the amiibo, as it needs to be registered to work

steps to reproduce:
1.get an amiibo .nbt file
2.go up to the amiibo box in the plaza
3.scan an amiibo
4.after talking to the amiibo, scan it again

Expected result: the amiibo is saved and allows gear to be obtained
Actual result: image
[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/8966048/yuzu_log.txt)

### LINKED ISSUE #8496: UI - Amibo Not Supported - Metroid Dread - Yuzu 1064
When attempting to add an Amibo to Metroid Dread I get the message Amibo Not Supported.

image

Log

[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/8976556/yuzu_log.txt)

### PR DESCRIPTION
**Pull request by german77**
Removes all hacks used to load an amiibo, adds an unique nfc interface for each npad. Implements functions needed for the cabinet applet. Reorganizes the service to match RE. Makes amiibo keys mandatory to read and write any data.

This fixes amiibo support on games like Shovel knight or erasing progress on SSBU. Only the cabinet applet is left to implement and a few details that I haven't fully understand.

fixes #8496
fixes #8493

### CODE REVIEW

File: src/common/input.h [hunk 1]
```diff
@@ -312,6 +332,14 @@ class OutputDevice {
     virtual CameraError SetCameraFormat([[maybe_unused]] CameraFormat camera_format) {
         return CameraError::NotSupported;
     }
+
+    virtual NfcState SupportsNfc() {
```
Comment (lioncash) [inline src/common/input.h:335]:
```suggestion
    virtual NfcState SupportsNfc() const {
```

File: src/core/hid/input_converter.cpp [hunk 2]
```diff
@@ -287,6 +287,20 @@ Common::Input::CameraStatus TransformToCamera(const Common::Input::CallbackStatu
     return camera;
 }
 
+Common::Input::NfcStatus TransformToNfc(const Common::Input::CallbackStatus& callback) {
+    Common::Input::NfcStatus nfc{};
+    switch (callback.type) {
+    case Common::Input::InputType::Nfc:
+        nfc = callback.nfc_status;
```
Comment (lioncash) [inline src/core/hid/input_converter.cpp:294]:
```suggestion
        return callback.nfc_status;
```

and the default case can just be `return {};`

File: src/core/hle/service/nfp/nfp_device.cpp [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+    Core::HID::ControllerUpdateCallback engine_callback{
+        .on_change = [this](Core::HID::ControllerTriggerType type) { NpadUpdate(type); },
+        .is_npad_service = false,
+    };
+    is_controller_set = true;
+    callback_key = npad_device->SetCallback(engine_callback);
+}
+
+NfpDevice::~NfpDevice() {
+    if (is_controller_set) {
```
Comment (lioncash) [inline src/core/hle/service/nfp/nfp_device.cpp:45]:
```suggestion
    if (!is_controller_set) {
        return;
    }
```

File: src/core/hle/service/nfp/nfp_device.cpp [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+
+    if (tag_data.settings.settings.appdata_initialized.Value() != 0) {
+        LOG_ERROR(Service_NFP, "Application area already exist");
+        return ApplicationAreaExist;
+    }
+
+    return RecreateApplicationArea(access_id, data);
+}
+
+Result NfpDevice::RecreateApplicationArea(u32 access_id, const std::vector<u8>& data) {
```
Comment (lioncash) [inline src/core/hle/service/nfp/nfp_device.cpp:473]:
```suggestion
Result NfpDevice::RecreateApplicationArea(u32 access_id, std::span<const u8> data) {
```

File: src/core/hle/service/nfp/nfp_device.cpp [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+    rng.GenerateRandomBytes(tag_data.application_area.data(), sizeof(ApplicationArea));
+    std::memcpy(tag_data.application_area.data(), data.data(), data.size());
+
+    tag_data.applicaton_write_counter++;
+    is_data_moddified = true;
+
+    return ResultSuccess;
+}
+
+Result NfpDevice::CreateApplicationArea(u32 access_id, const std::vector<u8>& data) {
```
Comment (lioncash) [inline src/core/hle/service/nfp/nfp_device.cpp:456]:
```suggestion
Result NfpDevice::CreateApplicationArea(u32 access_id, std::span<const u8> data) {
```

File: src/input_common/drivers/virtual_amiibo.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+namespace InputCommon {
+constexpr PadIdentifier identifier = {
+    .guid = Common::UUID{},
+    .port = 0,
+    .pad = 0,
+};
+
+VirtualAmiibo::VirtualAmiibo(std::string input_engine_) : InputEngine(std::move(input_engine_)) {}
+
+VirtualAmiibo::~VirtualAmiibo() {}
```
Comment (lioncash) [inline src/input_common/drivers/virtual_amiibo.cpp:23]:
```suggestion
VirtualAmiibo::~VirtualAmiibo() = default;
```

File: src/input_common/input_engine.cpp [hunk 7]
```diff
@@ -4,6 +4,7 @@
 #include "common/logging/log.h"
 #include "input_common/input_engine.h"
 
+#pragma optimize("", off)
```
Comment (lioncash) [inline src/input_common/input_engine.cpp]: This should be removed

File: src/input_common/input_engine.cpp [hunk 8]
```diff
@@ -404,3 +442,5 @@ void InputEngine::DeleteCallback(int key) {
 }
 
 } // namespace InputCommon
+
+#pragma optimize("", on)
```
Comment (lioncash) [inline src/input_common/input_engine.cpp]: Ditto

File: src/input_common/input_engine.h [hunk 9]
```diff
@@ -127,6 +128,17 @@ class InputEngine {
         return Common::Input::CameraError::NotSupported;
     }
 
+    // Request nfc data from a controller
+    virtual Common::Input::NfcState SupportsNfc([[maybe_unused]] const PadIdentifier& identifier) {
```
Comment (lioncash) [inline src/input_common/input_engine.h:132]:
```suggestion
    virtual Common::Input::NfcState SupportsNfc([[maybe_unused]] const PadIdentifier& identifier) const {
```

File: src/common/input.h [hunk 10]
```diff
@@ -312,6 +331,14 @@ class OutputDevice {
...
```
Comment (lioncash) [inline src/common/input.h:335]:
```suggestion
    virtual NfcState SupportsNfc() const {
```

**Review by liamwhite — APPROVED** [review]

**Review by bunnei — APPROVED** [review]

**Pull request merged**